Step4: 포커 딜러와 게임 테스트

kjoonk opened the pull request:
Step4 구현입니다. 클래스를 구현하면서 테스트도 같이 병행했습니다. 전보다 커밋을 잘게 자르려고 했습니다. 중간에 Step3 피드백을 반영한 커밋이 있습니다.

## Dealer
- 초반엔 게임을 준비하는 `prepare()`와 게임을 시작하는 `startGame()` 둘다 public으로 구현하였는데, `prepare()`를 `startGame()`에 포함시키고 `private`으로 전환했습니다.

## Player
- 참가자에 대한 간단한 클래스입니다.

## Playable
- `Dealer`를 구현하던 중, 카드를 돌리는 부분에서 `Player`와 같은 형태를 따라야 할 필요가 생겼습니다. 그래서 `Playable` 프로토콜을 간단히 구현하여 `Dealer`와 `Player`가 이를 준수하도록 하였습니다. 
- 공통 메소드인 `receive()`는 extension으로 구현했습니다.
- `cards` 속성은 딜러에게 받은 카드를 담는 변수입니다.

## Game
- 초반에는 게임 룰을 7-stud로 고정하여 구현했고, 나중에 `Rule`을 구현했습니다. `Game`에 `Rule`이 속하는 건 자연스러운데, 규칙에 따라 바뀌는 부분을 어디서 처리해야 할지 고민이었습니다. 결국 `Dealer`가 카드를 나누어주니 그쪽에서 처리하는 게 나아보였습니다.
- 나누어 주는 카드의 개수는 `enum Rule`의 `rawValue`로 처리했습니다. enum이 같은 종류의 상수를 묶는데 유용하여 자주 사용하는데, 좀더 욕심을 내면 생각처럼 잘 되지 않는 녀석입니다.
- 구현을 하던 도중, 초기화할 때 필요한 속성을 인자로 받으라는 피드백이 따올라 적용했습니다.
- 참가자 수를 벗어난 경우에 오류를 던지게 했습니다. 오류를 다루는 것이 익숙치 않아서 아직 손이 잘 가질 않습니다.

## CardGameFactory
- `Game`을 구현하던 중에, player를 내부에서 생성하다가 밖으로 빼내는 게 낫겠단 생각이 들었습니다. 그래서 기존 `DeckFactory`를 확장해 이름도 변경했습니다.

## 기타
- 피드백 중 테스트에 사용한 카드 수도 매직 넘버이니 바꾸라는 내용을 작업하는 와중에, 이미 만들어 놓은 `Card.Rank`와 `Card.Suit`을 활용할 생각을 못 하고 지나쳤다는 걸 알게 됐습니다.
- 이번 스텝에는 타이머를 사용했습니다. 순수 구현 시간이 2시간 35분으로 기록됐는데, 타이머를 재지 않은 시간도 있으니 실제로는 훨씬 깁니다.

kjoonk committed:
Dealer 클래스 생성

- 카드를 돌리는 빈 메소드 정의

kjoonk committed:
Dealer 일부 구현

- prepare 메소드 구현
- startGame 메소드 구현 중: 작동하지 않음. Player와 Dealer가 같은 프로토콜을 따르도록 할 필요가 생김

kjoonk committed:
Playable 프로토콜 선언

- Dealer, Player가 이 프로토콜을 준수하도록 변경

kjoonk committed:
Dealer, Playable 일부 기능 구현

- 구현을 위해 게임 규칙은 7-stud로 고정시킴. 추후 구현할 것
- prepare, startGame 메소드 구현
- give 메소드 구현
- Dealer 관련 테스트 작성

kjoonk committed:
PR 피드백 반영: DeckTests 변경

- DeckTests: 매직 넘버를 상수로 변경
- DeckFactory.createCards 메소드 private으로 변경

kjoonk committed:
Game 기본 구현

- 인원 수 제한

kjoonk committed:
CardGameFactory에 players 생성 메소드 추가

- DeckFactory를 CardGameFactory로 변경
- 관련 테스트 추가

kjoonk committed:
Game Rule 구현(7-stud, 5-stud)

- Game.Rule의 rawValue로 카드 수 결정
- CardGameFactory 파일 생성
- Dealer.prepare 메소드 private 처리
- 관련 테스트 추가

kjoonk committed:
일부 속성 private 처리

- CardGameFactory 일부 메소드 변경: 선택적으로 인자를 받을 수 있게 변경
- 상단 코멘트 제거

godrm commented on CardGameApp/CardGameApp/CardGameFactory.swift:
```diff
@@ -0,0 +1,28 @@
+        return Card.Rank.allCases.map {
+            Card(suit: suit, rank: $0)
+        }
+    }
```
이렇게 여러 객체를 생성하는 메소드를 모아놓은 팩토리 객체가 어떤 의도와 의미가 있는건가요?
static 함수로 선언하면 객체로 만들었을 때 장점이 없어, CardGameFactory에 대한 의존성을 줄일 방법이 없습니다.
각 객체로 분산시키는 게 어떨까요?

kjoonk replied:
각 클래스로 옮겼습니다. init()으로 추가하는 방법을 취했습니다.

godrm commented on CardGameApp/CardGameApp/Dealer.swift:
```diff
@@ -0,0 +1,52 @@
+    /// playerㅇ에게 한 장의 카드를 건넴
+    func give(card: Card, to player: Playable) {
+        var player = player
+        player.receive(card: card)
```
대상 객체도 있고 카드도 있는데 꼭 여기서 처리해야하는 걸까요? 
그냥 `player.receive(card: card)` 이걸 직접 처리해도 되지 않을까요? 
이 메소드가 어떤 의도인지 모르겠어요.

kjoonk replied:
receive()가 처리하고 give()는 제거했습니다.

godrm commented on CardGameApp/CardGameApp/Dealer.swift:
```diff
@@ -0,0 +1,52 @@
+}
+
+class Player: Playable {
+    var cards = [Card]()
```
참가자가 매우 수동적인 객체가 되었네요. 
private var 로 선언해보세요.

kjoonk replied:
private으로 처리하면서 Playable도 변경하게 되었습니다.

godrm commented on CardGameApp/CardGameApp/Dealer.swift:
```diff
@@ -0,0 +1,52 @@
+        let numberOfCards = rule.rawValue
+        (1...numberOfCards).forEach { _ in
+            dealCardsOfOneRound(players)
+        }
```
게임과 딜러, 덱 포함 관계가 궁금합니다. 
메소드 이름이나 동작을 보면 딜러가 최상위 객체인 건가요? 
enum 타입을 만들어서 사용할 때 rawValue 에 접근하지 않도록 하는 게 좋습니다. 
enum 타입 자체를 활용할 수 있는 게 좋습니다.

godrm commented on CardGameApp/CardGameApp/Dealer.swift:
```diff
@@ -0,0 +1,52 @@
+}
+
+protocol Playable {
+    var cards: [Card] { get set }
```
프로토콜은 속성을 추상화하기 보다는 메소드를 추상화하는 게 좋습니다.

kjoonk replied:
card 속성을 제거하고 메서드를 추가했습니다. 이 과정에서 extension도 없애고, 각 클래스에 구현했습니다. card를 private으로 선언하니 extension 내에서 사용할 수 없는 문제가 생겼습니다.
구현이 중복되어 있는데, 해결 방법을 찾지 못했습니다.

godrm commented on CardGameApp/CardGameApp/Dealer.swift:
```diff
@@ -0,0 +1,52 @@
+extension Playable {
+    mutating func receive(card: Card) {
+        self.cards.append(card)
+    }
```
모든 프로토콜 채택하는 객체가 사용할 default implementation 으로 구현할 꺼면 프로토콜에도 명시적으로 선언해주는 게 어떨까요?

kjoonk replied:
반영했습니다.

godrm commented on CardGameApp/CardGameApp/Game.swift:
```diff
@@ -0,0 +1,30 @@
+    }
+    
+    func start() {
+        dealer.startGame(rule: rule, players: players)
```
게임 구조체의 역할과 책임은 무엇일까요?
최상위 객체라서 하위에 있는 객체들 협력하도록 만들어 주는 것이라고 생각하는데,
지금 구조에서는 dealer가 모든걸 다하고 있는 것 같습니다.
startGame 내부에 있는 코드 중에서 일부는 여기, 일부는 dealer, 일부는 player가 해야할 역할이지 않을까 고민해보세요.

kjoonk replied:
살펴보니 Dealer에서 대부분 처리하고 있었습니다. 일부를 Game으로 옮겼습니다.

godrm reviewed (commented):
테스트 코드와 함께 로직을 만드는 경험은 매우 중요합니다.
덧붙여서 지금 구현한 팩토리 패턴에 대해 다시 한 번 고민해보세요.
다른 피드백도 확인해보세요.

kjoonk committed:
Dealer, Player의 cards 속성을 private으로 변경

- Playable 프로토콜 변경: isReadyToGame 메소드 추가
- Dealer.give 메소드 제거: receive 메소드에서 처리하게 변경
- 그에 따른 테스트 코드 수정

kjoonk committed:
Dealer.startGame 기능을 Game으로 옮김

- Dealer에게 넘겼던 기능을 Game이 하도록 함
- Dealer는 룰에 상관없이 한장씩 카드를 돌리도록 변경
- DealerTests에 중복되어 있던 Game 관련 테스트 제거

kjoonk committed:
Rule의 rawValue 사용하지 않도록 변경

- Rule을 Game 밖으로 빼냄

kjoonk committed:
Rule을 enum에서 클래스로 변경

- StudRule로 이름을 바꾸고 각 rule을 상속 받아 표현하도록 구현
- StudRule 별도 파일로 분리
- Player 별도 파일로 분리

kjoonk committed:
CardGameFactory 제거

- Factory에 있던 메서드를 각 클래스로 옮김
- 관련 테스트 변경
- PlayerTests 추가

kjoonk commented:
피드백 반영한 것 올렸습니다.

godrm commented on CardGameApp/CardGameApp/Game.swift:
```diff
@@ -20,11 +15,28 @@ struct Game {
+    init(rule: StudRule, numberOfPlayers: Int) throws {
+        do {
+            let players = try Player.create(count: numberOfPlayers)
+            try self.init(rule: rule, dealer: Dealer(), players: players)
```
rule은 이미 StudRule 타입이고 numberOfPlayers는 Int 네요. 
numberOfPlayers 도 Int가 아니라 더 구체적인 타입이었으면 좋았을 것 같습니다.
